Replace secrets, node acceptance, and CA hash with join tokens

aaronlehmann opened the pull request:
Implement the swarmkit parts of the flow described in https://github.com/docker/docker/issues/24430.

I need to add a few tests, but otherwise this PR is nearly ready.

cc @diogomonica @aluzzardi @stevvooe @tonistiigi @tiborvass

aaronlehmann commented:
Added a controlapi test to cover token rotation. Still need to add a test in the `ca` package that covers rotation.

stevvooe commented on api/ca.proto:
```diff
@@ -34,8 +34,12 @@ message NodeCertificateStatusResponse {
+
 	// Secret represents a user-provided string that is necessary for new
 	// nodes to join the cluster
 	string secret = 3;
```
Does this become token?

aaronlehmann commented:
Added a test in the `ca` package for token rotation. Removed WIP.

PTAL

dperny commented:
This will also require fixing the integration tests in docker/docker.

aaronlehmann commented:
Don't merge this until the docker/docker PR is ready.

tonistiigi commented on cmd/swarmd/main.go:
```diff
@@ -107,12 +107,7 @@ var (
-			}
-
-			secret, err := cmd.Flags().GetString("secret")
+			joinToken, err := cmd.Flags().GetString("secret")
```
s/secret/join-token/

diogomonica replied:
not secret anymore.

diogomonica commented on agent/node.go:
```diff
@@ -43,11 +43,8 @@ type NodeConfig struct {
-
-	// Secret to be used on the first certificate request.
-	Secret string
+	// JoinToken is the token to be used on the first certificate request.
```
Each node's first certificate request.

Maybe also add that the default implementation also includes the CA Hash?

diogomonica commented on api/ca.proto:
```diff
@@ -34,11 +34,15 @@ message NodeCertificateStatusResponse {
 message IssueNodeCertificateRequest {
-	NodeRole role = 1;
+	// DEPRECATED: Role is now selected based on which secret is matched.
+	NodeRole role = 1 [deprecated=true];
```
Why are we deprecating VS removing? Are we already on API freeze mode?

aaronlehmann replied:
I think we could remove this one.

For the stuff that persists in raft, I'd prefer to avoid messing up compatiblity, because it will mean anyone upgrading from a previous -rc will have a broken cluster with a bunch of confusing protobuf errors.

dperny replied:
Isn't it a thing with protobuf that you can drop number fields as long as you don't reuse numbers because when the newer version unmarshals older protobufs, it'll just ignore those fields? Does that not apply in this case?

aaronlehmann replied:
That's correct. The fields are left in as deprecated to avoid accidentally reusing the field numbers.

dperny replied:
Could also do, like, `reserved 1; // formerly NodeRole, now deprecated`

diogomonica commented on ca/config.go:
```diff
@@ -148,10 +154,36 @@ func NewConfigPaths(baseCertDir string) *SecurityConfigPaths {
+
+	var nn, digest big.Int
+	nn.SetBytes(secretBytes[:])
+	digest.SetString(rootCA.Digest.Hex(), 16)
```
Isn't this giving a token that has the CA hash as base16 instead of base36?

aaronlehmann replied:
No, it is parsing the hex representation of the hash to turn it into base36.

diogomonica commented on ca/config.go:
```diff
@@ -148,10 +154,36 @@ func NewConfigPaths(baseCertDir string) *SecurityConfigPaths {
+	var nn, digest big.Int
+	nn.SetBytes(secretBytes[:])
+	digest.SetString(rootCA.Digest.Hex(), 16)
+	return fmt.Sprintf("SWMTKN-1-%s-%0[2]*s", digest.Text(joinTokenBase), maxGeneratedSecretLength, nn.Text(joinTokenBase))
```
Why are we using `maxGeneratedSecretLength` instead of the actual len?

aaronlehmann replied:
We're zero-padding the secret to 128 bits. It's not really necessary, but I felt like this would be a good idea so the length of the secret doesn't leak the upper bits. Happy to change it.

tonistiigi replied:
Why pad only secret and not digest?

aaronlehmann replied:
I'll pad the digest as well.

diogomonica commented on ca/config.go:
```diff
@@ -148,10 +154,36 @@ func NewConfigPaths(baseCertDir string) *SecurityConfigPaths {
+	return fmt.Sprintf("SWMTKN-1-%s-%0[2]*s", digest.Text(joinTokenBase), maxGeneratedSecretLength, nn.Text(joinTokenBase))
+}
+
+func tokenCAHash(token string) (digest.Digest, error) {
```
`getCAHashFromToken`?

diogomonica commented on ca/server.go:
```diff
@@ -402,6 +340,13 @@ func (s *Server) Run(ctx context.Context) error {
 		state.EventUpdateCluster{},
 	)
+
+	// Do this after updateCluster has been called, so isRunning never
```
Don't understand how this is achieving that goal?

aaronlehmann replied:
When an RPC is invoked, it first checks `isRunning`. `isRunning` checks for `s.ctx` to determine whether RPCs can run yet. So it's important to set `s.ctx` only after we initialize the relevant variables.

diogomonica commented on ca/server_test.go:
```diff
@@ -343,160 +182,89 @@ func TestNewNodeCertificateRequiresSecret(t *testing.T) {
+		return store.UpdateCluster(tx, clusters[0])
+	}))
+
+	time.Sleep(500 * time.Millisecond)
```
lame

diogomonica commented:
LGTM

aaronlehmann commented:
Added a test `TestLoadOrCreateSecurityConfigWrongCAHash` that ensures that we properly check the CA cert hash part of the token.

tonistiigi commented:
@aaronlehmann 

Got this panic when shutting down a manager. I had done a rotation before that.

```
panic: could not find ':' in digest:

goroutine 84 [running]:
panic(0xee3c20, 0xc8203d7470)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/docker/swarmkit/vendor/github.com/docker/distribution/digest.Digest.sepIndex(0x0, 0x0, 0x6)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/vendor/github.com/docker/distribution/digest/digest.go:135 +0xda
github.com/docker/swarmkit/vendor/github.com/docker/distribution/digest.Digest.Hex(0x0, 0x0, 0x0, 0x0)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/vendor/github.com/docker/distribution/digest/digest.go:124 +0x37
github.com/docker/swarmkit/ca.GenerateJoinToken(0xc82028e4e0, 0x0, 0x0)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/ca/config.go:169 +0x28c
github.com/docker/swarmkit/manager.(*Manager).Run.func2.1(0x7fae4666bf70, 0xc8208d7830, 0x0, 0x0)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/manager/manager.go:318 +0x139
github.com/docker/swarmkit/manager/state/store.(*MemoryStore).update(0xc820276150, 0x7fae46669958, 0xc82008a480, 0xc82024ef18, 0x0, 0x0)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/manager/state/store/memory.go:228 +0x178
github.com/docker/swarmkit/manager/state/store.(*MemoryStore).Update(0xc820276150, 0xc82024ef18, 0x0, 0x0)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/manager/state/store/memory.go:270 +0x54
github.com/docker/swarmkit/manager.(*Manager).Run.func2(0xc8202e0cc0, 0xc820270100, 0x7fae46667518, 0xc82026ba80)
	/home/vagrant/data/go/src/github.com/docker/swarmkit/manager/manager.go:340 +0x48c
created by github.com/docker/swarmkit/manager.(*Manager).Run
	/home/vagrant/data/go/src/github.com/docker/swarmkit/manager/manager.go:440 +0x18f
```

Would be easier to test if swarmctl had a way to see/rotate tokens. Also I saw references to acceptance policies in swarmctl/api.

aaronlehmann commented:
I pushed a change that should fix the panic. The `RootCA` struct has a `Digest` field but it was not being set when we retrieved the cert from a remote CA. I think this field was not being used before this PR, so it was not being set consistently.

The acceptance policy stuff in api should be related to deprecated fields. I am removing the related items from swarmctl. I'll also add token display/rotation to swarmctl.

aaronlehmann commented:
Removed acceptance policy stuff from swarmctl, added join tokens to to the `swarmctl cluster inspect` output, and added a `--rotate-join-toker=(worker|manager)` flag to `swarmctl cluster update`.

tonistiigi commented:
LGTM

diogomonica commented:
@aaronlehmann failing CI on `TestManager`

aaronlehmann commented:
Looks unrelated - my latest changes were only to swarmctl. I started a new run.

aaronlehmann commented:
Tests passed this time.

diogomonica commented on ca/config.go:
```diff
@@ -28,16 +31,21 @@ const (
+
+	generatedSecretEntropyBytes = 16
+	joinTokenBase               = 36
+	// ceil(log(2^128-1, 36))
```
repeated comment?

aaronlehmann replied:
huh?
